lib/srv/db/postgres/redshift-activate-user.sql

```diff
@@ -0,0 +1,40 @@
+CREATE OR REPLACE PROCEDURE teleport_activate_user(username varchar, roles text)
+LANGUAGE plpgsql
+AS $$
+DECLARE
+    roles_length integer;
+    cur_roles_length integer;
+BEGIN
+    roles_length := JSON_ARRAY_LENGTH(roles);
+
+    -- If the user already exists and was provisioned by Teleport, reactivate
+    -- it, otherwise provision a new one.
+    IF EXISTS (SELECT user_id FROM svv_user_grants WHERE user_name = username AND admin_option = false AND role_name = 'teleport-auto-user') THEN
+        -- If the user has active connections, make sure the provided roles
+        -- match what the user currently has.
+        IF EXISTS (SELECT user_name FROM stv_sessions WHERE user_name = CONCAT('IAM:', username)) THEN
+          SELECT INTO cur_roles_length COUNT(role_name) FROM svv_user_grants WHERE user_name = username AND admin_option=false AND role_name != 'teleport-auto-user';
+          IF roles_length != cur_roles_length THEN
+            RAISE EXCEPTION 'User has active connections and roles have changed';
```

Comment:
Sounds good! Will update to use SQLSTATE once I rebase the master.

Reply:
Unfortunately, the Redshift RAISE command doesn't support setting SQLSTATE. I tried `SET SQLSTATE TO <>` but didn't work either. So I will keep it as it is.

> Following my comment above, if the procedures return the SQLSTATE, this assertion might be more accessible and less error-prone.

To add a return code, we will need to update the existing Postgres scripts as well,  which I would rather keep as it is 
 as well. What do you think?

Reply:
> Unfortunately, the Redshift RAISE command doesn't support setting SQLSTATE. I tried SET SQLSTATE TO <> but didn't work either. So I will keep it as it is.

Yeah, it seems only to support string. What do you think about adding the SQLSTATE into the error message (for example, `'TP002: User has active connections and roles have changed'`)? I know this is not ideal, but we would still report our code to the application and users.

> To add a return code, we will need to update the existing Postgres scripts as well, which I would rather keep as it is as well. What do you think?

We could do this update later, but having Redshift procedures already returning the SQLSTATE would make things easier.

Reply:
> What do you think about adding the SQLSTATE into the error message 

Added

> We could do this update later, but having Redshift procedures already returning the SQLSTATE would make things easier.

I tried to add an out state to both activate and deactivate functions but got the following error:
```
ERROR: Nested call of an SP cannot have OUT params (SQLSTATE 42601)]
``
meaning cannot add OUT to deactivation since it's called by activation. not sure if this is a Redshift limitation or also applies to Postgres. I will just keep them as it is. The `TPXXX` code the RAISE message should be strong enuf.